File: JavaScript/Poker.js
function run() {
    document.getElementById('result').innerHTML = '';
    var pokersArr = gameIn();
    var sum = getPokersSum(pokersArr);
    judge(pokersArr, sum);
}
function gameIn() {
    var toWrite = document.getElementById('result');
    var gameIn = document.getElementById('in').value;
    if (gameIn === '') {
        toWrite.innerHTML = 'empty pokers';
        return;
    }
    // delete wrong characters
    gameIn = gameIn.replace(/[^\dajqkAJQK]/g, '');
    // change to upper case
    gameIn = gameIn.toUpperCase();
    // take the first 5
    gameIn = gameIn.substr(0, 5);
    // less than 5 is not right
    if (gameIn.length < 5) {
        toWrite.innerHTML = 'not enough pokers';
        return;
    }
    // change to array
    gameIn = gameIn.split('');
    // change 1 to A, change 0 to 10
    for (var i = 0; i < 5; i++) {
        if (gameIn[i] === '1') {
            gameIn[i] = 'A';
        } else if (gameIn[i] === '0') {
            gameIn[i] = '10';
        }
    }
    return gameIn;
}
function judge(pokersArr, sum) {
    if (judgeBigBonus(pokersArr, sum) === false) {
        if (judgeOtherBonus(pokersArr, sum) === false) {
            var toWrite = document.getElementById('result');
            toWrite.innerHTML = 'LOSE';
        }
    }
}

function judgeBigBonus(pokersArr, sum) {
    var toWrite = document.getElementById('result');
    // bomb
    var pokerCount = [], arrPokers = [];
    for (var i = 0; i < 5; i++) {
        if (arrPokers.indexOf(pokersArr[i]) === -1) {
            arrPokers[arrPokers.length] = pokersArr[i];
            pokerCount[arrPokers.indexOf(pokersArr[i])] = 1;
        } else {
            pokerCount[arrPokers.indexOf(pokersArr[i])]++;
        }
    }
    if (pokerCount.indexOf(4) !== -1) {
        toWrite.innerHTML = 'BOMB : ' + arrPokers[pokerCount.indexOf(4)];
        return true;
    }
    // five bulls
    if (sum === 50) {
        toWrite.innerHTML = 'FIVE BULLS';
        return true;
    }
    // five small bulls
    if (sum === 10) {
        toWrite.innerHTML = 'FIVE SMALL BULLS';
        return true;
    }
    return false;
}

function judgeOtherBonus(pokersArr, sum) {
    sum = getPokersSum(pokersArr);
    value = sum % 10 === 0 ? 10 : sum % 10;
    arr2Calc = [];
    for (var i = 0; i < 5; i++) {
        arr2Calc[arr2Calc.length] = getPokerVal(pokersArr[i]);
    }
    if (judge10(arr2Calc, 0, 1, 2)) {
        showBull(pokersArr, value, 0, 1, 2);
        return true;
    } else if (judge10(arr2Calc, 0, 1, 3)) {
        showBull(pokersArr, value, 0, 1, 3);
        return true;
    } else if (judge10(arr2Calc, 0, 1, 4)) {
        showBull(pokersArr, value, 0, 1, 4);
        return true;
    } else if (judge10(arr2Calc, 0, 2, 3)) {
        showBull(pokersArr, value, 0, 2, 3);
        return true;
    } else if (judge10(arr2Calc, 0, 2, 4)) {
        showBull(pokersArr, value, 0, 2, 4);
        return true;
    } else if (judge10(arr2Calc, 0, 3, 4)) {
        showBull(pokersArr, value, 0, 3, 4);
        return true;
    } else if (judge10(arr2Calc, 1, 2, 3)) {
        showBull(pokersArr, value, 1, 2, 3);
    } else if (judge10(arr2Calc, 1, 2, 4)) {
        showBull(pokersArr, value, 1, 2, 4);
        return true;
    } else if (judge10(arr2Calc, 1, 3, 4)) {
        showBull(pokersArr, value, 1, 3, 4);
        return true;
    } else if (judge10(arr2Calc, 2, 3, 4)) {
        showBull(pokersArr, value, 2, 3, 4);
        return true;
    } else {
        return false;
    }
}
function getPokersSum(pokersArr) {
    var sum = 0;
    for (var i = 0; i < 5; i++) {
        sum += getPokerVal(pokersArr[i]) * 1;
    }
    return sum;
}

function getPokerVal(poker) {
    if (poker === 'A') {
        return 1;
    } else if (poker === 'J' || poker === 'Q' || poker === 'K') {
        return 10;
    } else {
        return poker;
    }
}
function judge10(arr2Calc, a, b, c) {
    return (arr2Calc[a] * 1 + arr2Calc[b] * 1 + arr2Calc[c] * 1) % 10 === 0;
}
function showBull(pokersArr, value, a, b, c) {
    var toWrite = document.getElementById('result');
    toWrite.innerHTML += 'BULLS ' + value + ':<br>';
    toWrite.innerHTML += '[ ';
    var d = -1, e;
    for (i = 0; i < 5; i++) {
        if (i === a || i === b || i === c) {
            toWrite.innerHTML += pokersArr[i] + ' ';
        } else if (d === -1) {
            d = i;
        } else {
            e = i;
        }
    }
    toWrite.innerHTML += '] + ( ' + pokersArr[d] + ' ' + pokersArr[e] + ' )';
}
function empty() {
    document.getElementById('in').value = '';
    document.getElementById('result').innerHTML = '';
}

File: Java/Poker.java

import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

class Poker {
    private String[] pokers;
    private HashMap<String, Integer> count;
    private int sum;
    private int value;

    Poker() {
        pokers = new String[5];
        count = new HashMap<>();
        sum = 0;
    }

    void gameStart() {
        if (gameIn()) {
            judge();
        }
    }

    private boolean gameIn() {
        System.out.println("what pokers ?");
        System.out.println("(\"0\" for 10, joker not allowed)");
        Scanner consoleIn = new Scanner(System.in);
        String in = consoleIn.nextLine();
        // change to upper case
        in = in.toUpperCase();
        // delete wrong characters
        in = in.replaceAll("[^\\dajqkAJQK]", "");
        // less than 5 is not right
        if (in.length() < 5) {
            System.out.println("not enough pokers");
            return false;
        }
        // take the first 5
        in = in.substring(0, 5);
        // change to array
        char[] rawPokers = in.toCharArray();
        // change 1 to A, change 0 to 10
        for (int i = 0; i < 5; i++) {
            if (rawPokers[i] == '1') {
                pokers[i] = "A";
            } else if (rawPokers[i] == '0') {
                pokers[i] = "10";
            } else {
                pokers[i] = String.valueOf(rawPokers[i]);
            }
        }
        // get count of each poker
        for (String item : pokers) {
            if (count.containsKey(item)) {
                count.put(item, count.get(item) + 1);
            } else {
                count.put(item, 1);
            }
        }
        // get sum of pokers
        for (String item : pokers) {
            sum += getPokerVal(item);
        }
//        {
//            // for debug
//            System.out.println(rawPokers);
//            System.out.println(sum);
//            System.out.println(count);
//            System.exit(0);
//        }
        return true;
    }

    private void judge() {
        if (!judgeBigBonus()) {
            if (!judgeOtherBonus()) {
                System.out.println("LOSE");
            }
        }
    }

    private boolean judgeBigBonus() {
        // bomb
        for (Map.Entry<String, Integer> entry : count.entrySet()) {
            if (entry.getValue() == 4) {
                System.out.printf("BOMB : %s\n", entry.getKey());
                return true;
            }
        }
        // five bulls
        if (sum == 50) {
            System.out.println("FIVE BULLS");
            return true;
        }
        // five small bulls
        if (sum == 10) {
            System.out.println("FIVE BULLS");
            return true;
        }
        return false;
    }

    private boolean judgeOtherBonus() {
        value = sum % 10 == 0 ? 10 : sum % 10;
        if (judge10(0, 1, 2)) {
            showBull(0, 1, 2);
            return true;
        } else if (judge10(0, 1, 3)) {
            showBull(0, 1, 3);
            return true;
        } else if (judge10(0, 1, 4)) {
            showBull(0, 1, 4);
            return true;
        } else if (judge10(0, 2, 3)) {
            showBull(0, 2, 3);
            return true;
        } else if (judge10(0, 2, 4)) {
            showBull(0, 2, 4);
            return true;
        } else if (judge10(0, 3, 4)) {
            showBull(0, 3, 4);
            return true;
        } else if (judge10(1, 2, 3)) {
            showBull(1, 2, 3);
            return true;
        } else if (judge10(1, 2, 4)) {
            showBull(1, 2, 4);
            return true;
        } else if (judge10(1, 3, 4)) {
            showBull(1, 3, 4);
            return true;
        } else if (judge10(2, 3, 4)) {
            showBull(2, 3, 4);
            return true;
        }
        return false;
    }

    private int getPokerVal(String poker) {
        switch (poker) {
            case "A":
                return 1;
            case "J":
            case "Q":
            case "K":
                return 10;
            default:
                return Integer.valueOf(poker);
        }
    }

    private boolean judge10(int a, int b, int c) {
        return (getPokerVal(pokers[a]) + getPokerVal(pokers[b]) + getPokerVal(pokers[c])) % 10 == 0;
    }

    private void showBull(int a, int b, int c) {
        System.out.printf("BULLS %d :\n[ ", value);
        int d = -1, e = -1;
        for (int i = 0; i < 5; i++) {
            if (i == a || i == b || i == c) {
                System.out.printf("%s ", pokers[i]);
            } else if (d == -1) {
                d = i;
            } else {
                e = i;
            }
        }
        System.out.printf("] + ( %s %s )\n", pokers[d], pokers[e]);
    }
}
